docs/content/sips/draft-000-templates.md

```diff
@@ -0,0 +1,365 @@
+    * Prompt for which language to use (of the ones available).
+    * Always prompt even if there is only one option, so the user knows if it's going to meet their needs.
+  * If there are NO templates named `foo` installed:
+    * If there are no templates at all installed, offer to install the default templates as above
+    * If there are templates installed:
+      * Message that `foo` is not installed
+      * List the templates that are installed, and point to `spin templates install`
+* If there IS a default language set, OR the user explicitly specifies a language:
+  * If there IS a template named `foo` in that language installed:
+    * Run it
+  * If there are templates named `foo` installed but not in the given language:
+    * Message that `foo` is not installed for your default language `Haskell`
+    * Prompt with the languages for which `foo` is installed
+  * If there are NO templates named `foo` installed:
+    * Same as if no default language
+
+(Note: if the user runs the command in silent/no-interaction mode, then instead
+of listing or prompting, we should just fail with an explanatory message.)
+
+### Applications and components
+
+Our current templates are of the "hello world" variety - they produce a small
+but complete application, including the application metadata and a starter
+component (code and wiring).
+
+It would be good to provide a way to add a component to an existing application.
+This is a little tricky because (a) it requires modification of an existing
+file, not just copying a file; and (b) the component templates have to be
+constrained by the trigger type of the application.
+
+@michelleN has also proposed that it should be possible to create _only_ the
+application metadata, up to and including the trigger type, but with no
+components.  (The `spin up` command would then be modified to detect if
+an app had not components and guide the user on how to add them.)  The
+user experience for this is a bit tricky.  One option is to have separate
+`spin init/spin new` verbs, but for some readers that will create ambiguity -
+I want to create an application, which one is right for me?  Another is
+to have flags on `spin new` to indicate "bare starter," "full starter,"
+or "incremental add," but this creates ceremony and cognitive load. A
+possible approach is to make it interactive - if no manifest exists,
+ask whether to create a bare one or one with a starter component.  But
+this may be a confusing question for users' first encounter with Spin!
+"I don't know, I just want to work!"
+
+_TODO: think about how bare and incremental `new` might impact the
+authoring of templates as well as the consumer experience._
+
+### Running a template without installing
+
+This could be something like:
+
+```
+spin new https://github.com/fermyon/spin/templates/http/rust
+```
+
+There would be no language option.
+
+We'd need to enforce that IDs couldn't look like URLs.
+
+### What's this 'default language' thing?
```

Comment:
Is a default language required? Seems like you could deploy e.g. bartholomew without caring what language it is written in.

Reply:
Oh that's interesting - you're thinking of templates with no source code, just references to existing Wasm modules?  (With the static files being the reason you need a separate application.)

I mean... no.  A default language isn't required.  It is a vague thought to save people having to type `-lang rust` EVERY. DAMN. TIME.  But yeah, if we did it, we would need to figure out how to make it interact with "no language" templates.  Thanks for the catch.

Reply:
I think the source of the confusion might be that I'm envisaging this as a filter/selector on templates that declare what language their code is in, and it's coming across as "we will convert the template content into Haskell for you using _checks notes_ Brian"

Is there a clearer term I could use for this?  Nothing concise is leaping to mind...

Reply:
Could be filtering on "tags", with special UI for tags that look like e.g. `lang:haskell`.

Reply:
Could be but I'm not sure what other use cases there would be for this _grin_.  Would you be able to expand on what you're thinking of? (or can jump on a call if you want to think out loud at me!)

Reply:
Wildly speculating: trigger type tags (`http`, `redis`), or a `cms` tag for Bartholomew et al.

Reply:
We probably need trigger type as a first class citizen anyway, so that when adding a component to an app it can offer compatible options.  Could be worth implementing it using a general mechanism for sure!

Reply:
One thing about encouraging authors to provide search keywords today, even if they're not terribly useful with the small number of templates we have today, is that if we set up a templates registry, it will be usefully searchable from day one - we won't need to bribe authors with cake to go back and add keywords to their templates.

Reply:
Especially if the UI can "nudge" template authors towards appropriate tags.

Reply:
> A default language isn't required. It is a vague thought to save people having to type -lang rust EVERY. DAMN. TIME.

Could we _just_ rely on template IDs and/or paths in a repo to identify templates, and leave grouping (whether to group them by tags, languages, or something else) to the template authors?